orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -370,17 +151,16 @@ void ReplaceNodeArgs(std::vector<Node*>& nodes,
   ReplaceNodeArgs(nodes, node_args, new_node_args);
 }
 
-std::string AddEventBeforeNode(
+Node& PrependEventNode(
```

Comment:
>PrependEventNode [](start = 6, length = 16)

would be nice to have some comments on what this function does

Reply:
Sure. Will do
```
// Create an event operator topologically before the input operator "node".
// All inputs of "node" would be re-wired to the passing-through outputs of the new event operator.
// That is,
//   upstream node -> node -> downstream node
// may become
//   upstream node -> event node (WaitEvent or RecordEvent) -> node -> downstream node
Node& PrependEventNode(
    Graph& graph,                                // Graph which contains "node" and the new event operator.
    Node* node,                                  // The anchor to prepend the new event operator.
    const std::string& event_op_type,            // Type of the new event operator, for example, "WaitEvent" or "RecordEvent".
    const std::string& event_op_name,            // Name's seed of the new event operator, for example, "WaitEvent" or "RecordEvent".
    const std::string& event_id_name,            // Name's seed of the event tensor consumed by the new event operator.
    std::vector<std::string>& new_input_names,   // Values to be added to input list of the transformed graph. Those values can be fed.
    std::vector<std::string>& new_output_names,  // Values to be added to output list of the transformed graph. Those values can be fatched.
    std::string& new_event_name,                 // Actual event name.
    std::string& new_output_name)               // First output of the created event operator.
```

---
In reply to: [476897719](https://github.com/microsoft/onnxruntime/pull/4916#discussion_r476897719) [](ancestors = 476897719)